frontends/p4-14/typecheck.cpp

```diff
@@ -411,6 +416,34 @@ class TypeCheck::AssignActionArgTypes : public Modifier {
     explicit AssignActionArgTypes(TypeCheck &s) : self(s) { setName("AssignActionArgTypes"); }
 };
 
+class TypeCheck::InsertImplicitCasts : public Transform, P4WriteContext {
+ public:
+    InsertImplicitCasts() { setName("InsertImplicitCasts"); }
+ private:
+    const IR::V1Program *global = nullptr;
+    bool        inAction = false;
+    profile_t init_apply(const IR::Node *root) override {
+        global = root->to<IR::V1Program>();
+        return Transform::init_apply(root); }
+    IR::Annotation *preorder(IR::Annotation *a) override { prune(); return a; }
+    IR::ActionFunction *preorder(IR::ActionFunction *af) override {
+        inAction = true;
+        visit(af->action);
+        inAction = false;
+        prune();
+        return af; }
+    IR::Expression *postorder(IR::Expression *op) override {
+        if (!inAction || isWrite()) return op;  // don't cast lvalues or refs outside actions
+        auto *type = inferTypeFromContext(getContext(), global);
+        if (type != IR::Type::Unknown::get() && !type->is<IR::Type_InfInt>() && type != op->type) {
+                LOG1("Need cast " << op->type << " -> " << type << " on " << op);
+                const IR::Expression *e = getOriginal<IR::Expression>();
+                if (*op != *e) e = op;  // undo Transform clone if it wasn't needed
```

Comment:
Strictly speaking it is not, as making an extra 1-node clone is pretty minor (the parent will always need cloning to reflect the new cast).  But the idea here is to insert a new (cast) node into the dag without modifying the node being cast if not otherwise needed.

That's why I said this would be a good candidate for a transform that does not autoclone (some discussions I've been having with Cole).